### DISCUSSION TITLE
ENMF shall compute over batch of user instead of full users

### DISCUSSION THREAD

**Discussion by rowedenny** [RUCAIBox/RecBole#999] [Q&A] [answered]
https://github.com/RUCAIBox/RecBole/blob/f56bcd001074b0a0f1dec5dd4ece90643d60ec8c/recbole/model/general_recommender/enmf.py#L85

Refer to the ENMF paper, the loss always goes with a batch of users. The `score` from function `forward` comes from users in current `interaction`, I assume it would be the same when computing `user_sum`, not all users.

**Comment by linzihan-backforward** [comment]
@rowedenny Thanks for your attention ! ENMF is a sepcific model that different from other gereral recommender models. 
Its loss  contains two part, the first part is calculated on all user and all item, and the second part is calculated on batch user. 
You can refer to the Tensorflow source code of ENMF.
https://github.com/chenchongthu/ENMF/blob/a3b186afca8f5ee36cf893734b4bbb55faa55a7a/code/ENMF.py#L98

**Comment by rowedenny** [comment]
Hi, thanks for your quick response. However, I am afraid that I cannot agree with you. We can investigate it from two aspects:
1. In ENMF paper, we can see clues from Equation 11. The loss consists of two terms, one comes from observed data, and the other one comes from all data. Note that here, both terms are summed over user `u` in a batch of users. So I assume they shall be consistent, i.e., both are computed over a batch of users.
![image](https://user-images.githubusercontent.com/29630980/137317377-2c212ad7-300d-4452-8a4c-6299a3796996.png)

2. Also notice in the TF code, `self.uid` actually comes from `self.uid = tf.nn.embedding_lookup(self.uidW, self.input_u)` https://github.com/chenchongthu/ENMF/blob/a3b186afca8f5ee36cf893734b4bbb55faa55a7a/code/ENMF.py#L77. 
In addition, the placehold `self.input_u` is fed with a batch of users, please see, https://github.com/chenchongthu/ENMF/blob/a3b186afca8f5ee36cf893734b4bbb55faa55a7a/code/ENMF.py#L254.

From these two aspects, I believe the current implementation has a minor issue, though I am not sure if one term is computed for full batch, while the other for a batch could result in loss inblance. Correct me if I have any misunderstandings.

**Comment by linzihan-backforward** [comment, marked as answer]
Hi @rowedenny. After consideration, i think you are right. It is exactly a BUG of our implementation. We will fix it as soon as possiable. Thank again for your carefully review of the code.

**Comment by rowedenny** [comment]
If my understanding is correct, we only need to update the `user_sum` by computing it over the batch of users from `interactions`, other things should be all the same.

Is that correct?

**Comment by rowedenny** [comment]
By the way, I may also point out one minor issue 
https://github.com/RUCAIBox/RecBole/blob/f56bcd001074b0a0f1dec5dd4ece90643d60ec8c/recbole/model/general_recommender/enmf.py#L49

Recall the padding item is zeros, however, the xavier_uniform_initialization initialization initializes all the embeddings, including the `padding_idx`. Therefore when we compute the loss with respect to L_A, this non-zero vector will also contribute to loss unexpectedly in the following places:
https://github.com/RUCAIBox/RecBole/blob/f56bcd001074b0a0f1dec5dd4ece90643d60ec8c/recbole/model/general_recommender/enmf.py#L68
https://github.com/RUCAIBox/RecBole/blob/f56bcd001074b0a0f1dec5dd4ece90643d60ec8c/recbole/model/general_recommender/enmf.py#L81
Unless we use a mask to mask out the padding influence. 

In short, I assume it is more appropriate to remove the xavier_normal_initialization initialization.
